decode/src/utils.rs

```diff
@@ -0,0 +1,231 @@
+        variants = { $( $bits:literal, $variant:ident, $display:literal $(,)? ),+ }
+        $(,)?
+    ) => {
+        // Sanity checks on the bit ranges.
+        static_assertions::const_assert!($high_bit < 8);
+        static_assertions::const_assert!($low_bit < 8);
+        static_assertions::const_assert!($low_bit <= $high_bit);
+
+        // Sanity check that the bit patterns are all within the mask.
+        $( static_assertions::const_assert_eq!($name::MASK | $bits, $name::MASK); )+
+
+        // Sanity check that the mask is _equal_ to all the bit patterns OR'd
+        // together.
+        static_assertions::const_assert_eq!( $name::MASK, $( $bits )|+ );
+
+        impl $name {
+            #[allow(dead_code)]
+            pub const HIGH_BIT: u8 = $high_bit;
+            #[allow(dead_code)]
+            pub const LOW_BIT: u8 = $low_bit;
+            pub const MASK: u8 = (0xff << $low_bit) & (0xff >> (7 - $high_bit));
+        }
+
+        #[doc = $docstring]
+        #[derive(Clone, Copy, Debug, PartialEq)]
+        #[cfg_attr(
+            any(feature = "api-traits", test),
+            derive(schemars::JsonSchema, serde::Deserialize, serde::Serialize)
+        )]
+        pub enum $name {
+            $($variant),+
+        }
+
+        impl ::core::fmt::Display for $name {
+            fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
+                use $name::*;
+                #[deny(overlapping_range_endpoints)]
+                match self {
+                    $( $variant => write!(f, "{}", $display), )+
+                }
+            }
+        }
+
+        impl ::core::convert::TryFrom<u8> for $name {
+            type Error = Error;
+
+            fn try_from(x: u8) -> Result<Self, Self::Error> {
+                use $name::*;
+                #[deny(overlapping_range_endpoints)]
+                match (x & Self::MASK) >> $low_bit {
+                    $( $bits => Ok($variant), )+
+                    _ => Err(Error::InvalidBitField),
+                }
+            }
+        }
+
+        impl ::core::convert::From<$name> for u8 {
+            fn from(x: $name) -> u8 {
+                use $name::*;
+                #[deny(overlapping_range_endpoints)]
+                match x {
+                    $( $variant => $bits, )+
+                }
+            }
+        }
+    };
+}
+
+#[cfg(test)]
+mod tests {
+    use super::extract_bit;
+
+    #[test]
+    fn test_extract_bit() {
+        for shift in 0..8 {
+            let expected = 1 << shift;
+            assert_eq!(extract_bit(expected, shift).unwrap(), true);
+            for other_shift in 0..8 {
+                if shift == other_shift {
+                    continue;
+                }
+                assert_eq!(extract_bit(expected, other_shift).unwrap(), false);
+            }
+        }
+
+        for shift in 8..=255 {
+            assert!(extract_bit(0, shift).is_err());
+        }
```

Comment:
I think the first loop could have been `0..=7` too. I did the last loop as inclusive because `shift` is a `u8`. If you write `for shift in 8..256`, then you get a compiler error about an overflowing literal. So I think the only way to write this is `8..` or `8..=255`.